cmd/katapult/console/selector.go

```diff
@@ -0,0 +1,491 @@
+package console
+
+import (
+	"container/list"
+	"io"
+	"math"
+	"os"
+	"strings"
+
+	"github.com/krystal/katapult-cli/internal/keystrokes"
+
+	"github.com/buger/goterm"
+	"golang.org/x/term"
+)
+
+const (
+	// Clarification string for a single item.
+	clarificationStringSingle = " (Press ENTER to make your selection): "
+
+	// Clarification string for multiple items.
+	clarificationStringMultiple = " (Press ENTER to select items and ESC when you are done with your selections): "
+)
+
+// TerminalInterface defines a interface for a compatible terminal. Used for unit testing.
+type TerminalInterface interface {
+	Height() int
+	Width() int
+	Print(items ...interface{}) (int, error)
+	Println(items ...interface{}) (int, error)
+	Clear()
+	Flush()
+	SignalInterrupt()
+	MakeRaw() (*term.State, error)
+}
+
+func intMin(x, y int) int {
+	if x > y {
+		return y
+	}
+	return x
+}
+
+func exactCompare(a, b interface{}) bool {
+	switch x := a.(type) {
+	case []string:
+		// The logic behind this: A string slice is just a fancy pointer with some extra data.
+		// If we compare the pointers, we can tell if they are equal.
+		return &x[0] == &b.([]string)[0]
```

Comment:
This feels very panicky. Are we sure zero length arrays won't be passed in, and that B will always be the same type?

Reply:
Yes. It is only called when items are selected, which would mean items exist in the first place. 👍